dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/eventdatavalue/DefaultEventDataValueService.java

```diff
@@ -0,0 +1,389 @@
+        }
+
+        DataElement dataElement = dataElementService.getDataElement( eventDataValue.getDataElement() );
+        if ( dataElement == null ) {
+            throw new  IllegalQueryException( "Given data element (" +  eventDataValue.getDataElement() + ") does not exist" );
+        }
+
+        createAndAddAudit( eventDataValue, dataElement, programStageInstance, AuditType.DELETE );
+        handleFileDataValueDelete( eventDataValue, dataElement );
+        programStageInstance.getEventDataValues().remove( eventDataValue );
+        programStageInstanceService.updateProgramStageInstance( programStageInstance );
+    }
+
+    @Override public void deleteEventDataValues( ProgramStageInstance programStageInstance, Set<EventDataValue> eventDataValues )
+    {
+        //TODO: See the Todo on the top of the class (Can be done with use of delete (-) parameter)
+
+        for ( EventDataValue eventDataValue : eventDataValues ) {
+            if ( StringUtils.isEmpty( eventDataValue.getDataElement() ) )
+            {
+                throw new  IllegalQueryException( "Data element is null or empty" );
+            }
+
+            DataElement dataElement = dataElementService.getDataElement( eventDataValue.getDataElement() );
+            if ( dataElement == null ) {
+                throw new  IllegalQueryException( "Given data element (" +  eventDataValue.getDataElement() + ") does not exist" );
+            }
+        }
+
+        for ( EventDataValue eventDataValue : eventDataValues ) {
+            DataElement dataElement = dataElementService.getDataElement( eventDataValue.getDataElement() );
+            createAndAddAudit( eventDataValue, dataElement, programStageInstance, AuditType.DELETE );
+            handleFileDataValueDelete( eventDataValue, dataElement );
+        }
+
+        programStageInstance.getEventDataValues().removeAll( eventDataValues );
+        programStageInstanceService.updateProgramStageInstance( programStageInstance );
+    }
+
+    // -------------------------------------------------------------------------
+    // Supportive methods
+    // -------------------------------------------------------------------------
+
+    private Set<EventDataValue> filterOutEmptyDataValues( Set<EventDataValue> eventDataValues ) {
+        return eventDataValues.stream().filter( dv -> !StringUtils.isEmpty( dv.getValue() ) ).collect( Collectors.toSet());
+    }
+
+    private String validateEventDataValue( EventDataValue eventDataValue ) {
+
+        if ( StringUtils.isEmpty( eventDataValue.getStoredBy() ) )
+        {
+            eventDataValue.setStoredBy( currentUserService.getCurrentUsername() );
+        }
+
+        if ( StringUtils.isEmpty( eventDataValue.getDataElement() ) )
+        {
+            return "Data element is null or empty";
+        }
+
+        DataElement dataElement = dataElementService.getDataElement( eventDataValue.getDataElement() );
```

Comment:
Can we use some kind of cache here? have a look at how its handled in AbstractEventService

Reply:
Correct. I overlooked it. Added.